Add shared MQTT client and ResourceManagement-based MQTT integration service

MathoMathiasCamara opened the pull request:
The following [Issue](https://github.com/PHOENIXCONTACT/MORYX-Framework/issues/660) is resolved in this PR.

# Motivation :
Based on https://github.com/PHOENIXCONTACT/MORYX-Framework/issues/656 and https://github.com/PHOENIXCONTACT/MORYX-Framework/issues/657, we want to allow applications to share resources throughout multiple MORYX instances. 

 This PR contains the following features:
- New feautures :
     - A shared MQTT client called `IManagedMqttClient`
     - `ResourceSynchronizationService` (Listen to resource change event and sends a message to the broker)
 - Old feature that were refactored to work with the new `IManagedMqttClient`
     - An `MqttEndpointService`  for  topic template support ex: "my-topic/{id:int}/invoke/{methodName}"
     - A `ResourceEventService` (listen to public Resource events and sends a message to the broker)
     - A `ResourceRpcEndpoint` (when the client receives a method invocation message, the method is invoked on the matching resource , any return value (except `Task<T>`) is sent back to the broker)

 # How the new feature works ?
 
## Mqtt Service

`Config/Moryx.Mqtt.Config.json`
```json
  {
    "Host": "localhost",
    "Port": 1883,
    "Id": "device-2",
    "Username": null,
    "Password": null,
    "Tls": false,
    "QoS": 1,
    "ReconnectDelayMs": 30000,
    "ReconnectWithClientSession": true,
    "ConfigState": "Generated",
    "RootTopic":  "moryx/device-2"
  }
```
```cs
// BlablaService.cs
 public class BlablaService : IMqttService
 {
    private IMyDependency _dependency;
    private IManagedMqttClient _client;
    public Blabla(IMyDependency dependency, IManagedMqttClient client)
    {
        _dependency = dependency;
        _client = client;
    }

    // </inherited>
    public override Task StartAsync(CancellationToken cancellationToken)
    {
        // your logic here
        // _client.OnConnectedAync += ....
        // _client.OnMessagereceived += ...
        return Task.CompletedTask;
    }

    // </inherited>
    public override Task StopAsync(CancellationToken cancellationToken)
    {
        //your logic here
        return Task.CompletedTask;
    }
 }
 

 // Program.cs

  builder.Services.AddMqttClient("Moryx.Mqtt.Config")
      .AddMqttService<BlablaService>()

```

## Endpoint feature
Example 1 :
```c#
// Program.cs
  builder.Services.AddMqttClient()
     .AddMqttEndpoints(builder =>
{
    builder.MapGet("my-topic/{id:int}", ctx =>
    {
         var messageBuilder = new MqttApplicationMessageBuilder();
        messageBuilder.WithTopic("another-topic");
        messageBuilder.WithPayload(MqttMessageSerialization.GetJsonPayload("Hello World!!!!"));
        return messageBuilder.Build();
    });
});
``` 

Example 2:
```c#
// Program.cs
  builder.Services.AddMqttClient()
     .AddMqttEndpoints();

// BlablaEndpoint.cs
 public class Blabla : IMqttEndpoint 
 {
    private IMyDependency _dependency;
    public Blabla(IMyDependency dependency)
    {
        _dependency = dependency;
    }

    public void Map(IMqttRouteBuilder route)
    {
        //topic template is supported here
        route.MapGet("my-topic/{id:int}/method/{methodName}", context =>
        {
             var messageBuilder = new MqttApplicationMessageBuilder();
            messageBuilder.WithTopic("another-topic");
            messageBuilder.WithPayload(MqttMessageSerialization.GetJsonPayload("Hello World!!!!"));
            return messageBuilder.Build();
        });
    }
}
```

1nf0rmagician commented on Directory.Build.targets:
```diff
@@ -30,11 +30,17 @@
-    <FrameworkReference Update="Microsoft.AspNetCore.App" Version="$(dotnetVersion)"/>
+    <PackageReference Update="MQTTnet" Version="4.3.1.873"/>
+    <PackageReference Update="MQTTnet.Extensions.ManagedClient"   Version="4.3.1.873"/>
+    <!--<PackageReference Update="MQTTnet.AspNetCore"  Version="4.3.1.873"/>-->
```
Please remove commented reference 😊

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Endpoints/README.md:
```diff
@@ -0,0 +1,8 @@
+ 1. `MqttApplicationMessage.ResponseTopic` if the  request (message) has a response topic.
+ 2. `resources/{identifier}/invoked/{methodName}` when the request (Message) doesn't have a ResponseTopic
+
+When a client sends a message to the following topic : `"resources/{identifier}/invoke/{methodName}"` , for example `resources/1236545/invoke/ChangeCapabilities`. The `ResourceRpcEndpoint`will be triggered and the matching Resource with the `Identifier` or `Name` `1236545` and the method that matches the name `ChangeCapabilities` will be invoked.
```
How to method parameters work, do you have any information about that? 🤔

MathoMathiasCamara replied:
I will add example.

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Endpoints/ResourceRpcEndpoint.cs:
```diff
@@ -0,0 +1,99 @@
+            var resource = resourceManagement.GetResource<IResource>(x =>
+                x is IIdentifiableObject obj && obj.Identity.Identifier == identifier
+                || x.Name == identifier);
+            return FindAndInvokeMethod(context, messageBuilder, identifier, methodName, resource);
```
Please add early returns if the method is null or empty and/or if the resource was not found 😊

MathoMathiasCamara replied:
got it.

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Endpoints/ResourceRpcEndpoint.cs:
```diff
@@ -0,0 +1,99 @@
+using Moryx.Serialization;
+using MQTTnet;
+
+namespace Moryx.Communication.Mqtt.Endpoints;
```
Open Question: Should we rather place all the endpoint in this namespace (at least that is how I understand your intend) or should we rather put the `ResourceRpcEndpoint` into the `Moryx.AbstractionLayer.Resources.Endpoints` package? 🤔

dbeuchler replied:
Its a special case.. this Endpoint has a dependency to mqtt stuff, and I definitely don't want a reference to that in `Moryx.AbstractionLayer.Resources.Endpoints`. For my POV, this endpoint should be separated because I also dont see a reference of `Moryx.AbstractionLayer` in `Moryx.Communication.Mqtt`

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Endpoints/ResourceRpcEndpoint.cs:
```diff
@@ -0,0 +1,99 @@
+            return null;
+        }
+
+        var methodToInvokeParameters = methodToInvoke.GetParameters();
```
You could split the parameter parsing into a separate private method to improve readability and maybe return early if a parameter could not be properly parsed 🤔

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Endpoints/ResourceRpcEndpoint.cs:
```diff
@@ -0,0 +1,99 @@
+                        : context.RequestMessage.ResponseTopic;
+        responseBuilder.WithTopic(topic);
+        var payload = methodResult ?? new { };
+        var jsonPayload = MqttMessageSerialization.GetJsonPayload(new MethodInvocationResult
```
We should document somewhere that this serialization is used and how it works for the `ResourceRpcEndpoint` 😊

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Endpoints/ResourceRpcEndpoint.cs:
```diff
@@ -0,0 +1,99 @@
+        var jsonPayload = MqttMessageSerialization.GetJsonPayload(new MethodInvocationResult
+        {
+            Value = payload,
+            ValueType = EntryConvert.TransformType(methodToInvoke.ReturnType)
```
Since there is an `EntryValueType` for exceptions you could properly return from all places where exceptions where thrown above 😊

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Endpoints/Moryx.Communication.Mqtt.Endpoints.csproj:
```diff
@@ -0,0 +1,13 @@
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
```
I like that this will actually be the first package from us with a readme on nuget 👍 Make sure to add the following here to have it packed

```xml
<PropertyGroup>
  <PackageReadmeFile>README.md</PackageReadmeFile>
</PropertyGroup>

<ItemGroup>
  <None Include="README.md" Pack="true" PackagePath="" />
</ItemGroup>
```

MathoMathiasCamara commented on src/Moryx.Communication.Mqtt.Endpoints/README.md:
i don't really get what you mean by "link to the full documentation". This is very a specific use case for the Mqtt Endpoint, showcasing how to use the `IMqttEndpoint` to create your own endpoint.

MathoMathiasCamara commented on src/Moryx.Communication.Mqtt.Endpoints/ResourceRpcEndpoint.cs:
```diff
@@ -0,0 +1,99 @@
+                // get the values of the method's parameters from the body of the message
+                parameters[i] = context.FromBody(methodToInvokeParameters[i].Name!, methodToInvokeParameters[i].ParameterType);
+            }
+            catch (Exception ex)
```
I will try a specific one.

MathoMathiasCamara commented on src/Moryx.Communication.Mqtt.Endpoints/ResourceRpcEndpoint.cs:
```diff
@@ -0,0 +1,99 @@
+                continue;
+            }
+        }
+        var methodResult = methodToInvoke.Invoke(resource, parameters);
```
But if the method itself has an exception, i will add a catch for that case.

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Endpoints/ResourceRpcEndpoint.cs:
```diff
@@ -0,0 +1,133 @@
+    MqttClientUserOptions options)
+    : IMqttEndpoint
+{
+    private const string message = "No resource or method found with the given Identifier:`{id}`, method:`{method}`. Task and Task<T> are not supported return types.";
```
I would prefer having dedicated error(message)s. You have three different errors here
- Resource not found
- Method found, but has unsupported return type Task
- Method not found on resource

Combined with early returns that would further improve the map method

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Services/ResourceEventService.cs:
```diff
@@ -0,0 +1,105 @@
+
+    private void FacadeStateChanged(object? sender, bool activated)
+    {
+        if (sender is ILifeCycleBoundFacade && activated)
```
`sender is ILifeCycleBoundFacade` is redundant

MathoMathiasCamara commented on src/Moryx.Communication.Mqtt/AttributeRouting/IMqttRouteTable.cs:
```diff
@@ -0,0 +1,23 @@
+
+// Modifications Copyright (c) Atlas Lift Tech Inc. All rights reserved.
+
+// Source https://github.com/Atlas-LiftTech/MQTTnet.AspNetCore.AttributeRouting/blob/master/Source/Routing
```
can you tell me how to get a version when there is no tag and no release only one branch (master) ? do you mean commit or what?

seveneleven replied:
Yes

dbeuchler commented on src/Moryx.Communication.Mqtt/Moryx.Communication.Mqtt.csproj:
```diff
@@ -0,0 +1,28 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Moryx.AbstractionLayer\Moryx.AbstractionLayer.csproj" />
```
`Moryx.Communication.Mqtt` should not reference `Moryx.AbstractionLayer` nor `Moryx.Runtime`

dbeuchler replied:
Yes, and thats the reason it must be separated from Moryx.Communication.Mqtt

dbeuchler commented on src/Moryx.Communication.Mqtt/README.md:
```diff
@@ -0,0 +1,265 @@
+- {name: long}
+- {name: long?}
+- {name?}
+- {name/*}
```
Where do they come from? Maybe reference another doc instead of copy the stuff into these docs.

1nf0rmagician commented on src/Moryx.Communication.Mqtt.Services/ResourceEventService.cs:
```diff
@@ -0,0 +1,126 @@
+
+    private void FacadeStateChanged(object? sender, bool activated)
+    {
+        if (activated)
```
As soon as a single restart happens during runtime you will have 2 subscriptions each. You also need to unsubscribe here 🙂  
_Porbably also true for the other services?_

MathoMathiasCamara commented on src/Moryx.Communication.Mqtt.Endpoints/README.md:
```diff
@@ -0,0 +1,84 @@
+# [ResourceRpcEndpoint](../../docs/articles/Mqtt/ResourceRpcEndpoint.md)
```
Does that mean all relative links currently in `AbstractionLayer.md` don't work ?

1nf0rmagician commented on src/Moryx.AbstractionLayer/Resources/Serializations/ResourceToJsonConverter.cs:
```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2025, Phoenix Contact GmbH & Co. KG
+// Licensed under the Apache License, Version 2.0
+#if NET8_0_OR_GREATER
```
Shouldn't be necessary here, `Moryx.AbstractionLayer` only targets net10

1nf0rmagician commented on src/Moryx.Resources.Management/Resources/ResourceManager.cs:
```diff
@@ -371,10 +370,6 @@ public async Task SaveAsync(Resource resource)
             foreach (var instance in newResources)
                 await AddResource(instance, true);
 
-            if (!isNew)
```
Did you remove this intentionally?

1nf0rmagician commented on src/Moryx.Communication.Mqtt/Moryx.Communication.Mqtt.csproj:
```diff
@@ -0,0 +1,28 @@
+    <TargetFramework>net10.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <PackageReadmeFile>README.md</PackageReadmeFile>
```
There is no file called Readme.md anymore

1nf0rmagician reviewed (commented):
I think we've gotten quite far now. Only a few remarks left. Also pls make sure to cleanup the history, i.e. squash commits to a set (or one) with meaningful commit messages.

Also check there are still build errors

1nf0rmagician commented on docs/articles/Mqtt/Moryx.Communication.Mqtt.md:
```diff
@@ -0,0 +1,167 @@
+# The goal:
```
Duplicated in [‎docs/articles/Mqtt/ManagedMqtt.md](https://github.com/PHOENIXCONTACT/MORYX-Framework/pull/720/changes#diff-3c94bf16b28052b13d88fb42158912461b02f3d98a4c3ee4247c658d2ac6b224)

1nf0rmagician commented on src/Moryx.Communication.Mqtt/Builders/MqttBuilder.cs:
```diff
@@ -0,0 +1,61 @@
+    public IMqttBuilder AddMqttEndpoints(Action<MqttRouteBuilder>? builder = null)
+    {
+        // find all endpoints via reflection
+        var endpoints = AppDomain.CurrentDomain.GetAssemblies()
```
FYI: The ReflectionTool has a method to get all public classes of a type

1nf0rmagician commented on src/Moryx.Resources.Management/Resources/ResourceManager.cs:
```diff
@@ -285,6 +285,7 @@ private void UnregisterEvents(Resource instance)
     private void OnResourceChanged(object sender, EventArgs eventArgs)
     {
         _ = Task.Run(() => SaveAsync((Resource)sender));
+        RaiseResourceChanged((IResource)sender);
```
Why do you need this change? Does it cause an infinte event loop?

1nf0rmagician commented on src/Moryx/Collections/WriteOnlyHashStore.cs:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    //<inherit>
+    public void Dispose()
```
Github warnings

1nf0rmagician commented on src/Moryx.Communication.Mqtt/Moryx.Communication.Mqtt.csproj:
```diff
@@ -0,0 +1,26 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
```
I think now you just removed the project readme, but the file at the bottom is still referenced and i still think it is a good idea to have the project readme. My prior comment just meant that the name does not match 😅

MathoMathiasCamara replied:
the name does not match and the file doesn't exist anymore. And Dennis also didn't like it.

MathoMathiasCamara commented on src/Moryx.Communication.Mqtt.Services/ResourceSynchronizationService.cs:
```diff
@@ -0,0 +1,252 @@
+
+    private void ResourceAdded(object? sender, IResource e)
+    {
+        //TODO: how to handle this case ?
```
i don't have to do that. i am already listening to the facade. so anything that happens to any resources, i will be notified.

 the question is:
Do we send this data to the broker? if yes how should we handle the cases (Resource Type is not present on receiver, or receiver csharp class is different from incoming, should we include the charp class information in the data, etc...)

1nf0rmagician replied:
I suggest we include the type information in the data, and if the type is not present at the target, we cannot create it and log an error on the instance not able to create the resource. What do you think?

MathoMathiasCamara replied:
that's a good place to start. i will use that an find a better alternative

MathoMathiasCamara commented on src/Moryx.Communication.Mqtt.Services/ResourceSynchronizationService.cs:
```diff
@@ -0,0 +1,252 @@
+
+    private void ResourceRemoved(object? sender, IResource e)
+    {
+        //TODO: how to handle this case ?
```
same as above. i don't have to do that.

MathoMathiasCamara commented:
- Added Unit test

MathoMathiasCamara commented:
- Added support for Delete and Add events
-  Added a new property `SynchronizationTypeId` to the `ResourceSynchronizationAttribute` to provide a balance between flexibility and type safety while handling assembly version mismatches gracefully across different MORYX instances and versions, to properly handle Add and delete.
